File: neural/lif.m
% Leaky Integrate-and-fire (IAF) neuron
clear; clf;

% parameters
tau_inv=.1; % inverse time constant
uu=0; % initial membrane voltage
I_ext=12; % constant external input
tspan=[0 100] % integration interval
theta=10; % firing threshold

% Integration with Euler method
t_step=0; for it=0:100;
    t_step=t_step+1;
    x=uu<theta;
    uu=x*(1-tau_inv)*uu+tau_inv*Iext;%+randn;
    u(t_step)=uu;
    t(t_step)=it;
    s(t_step)=1-x;
end

subplot('position',[0.13 0.13 1-0.26 0.6])
    plot(t,u);
    hold on; plt([0 100], [10 10], '--'');
    axis([0 100 0. 12]);
    xlabel('time [\tau]');
    ylabel('u(t)');

subplot('position',[0.13 0.8 1-0.26 0.1])
    plot(t,s,'.', 'markersize', 20);
    axis([0 100 0.5 1.5]);
    set(gca, 'xtick', [], 'ytick', []);
    ylabel('spikes');

File: neural/lif.py
import matplotlib.pyplot as plt
import numpy as np

'''
Leaky integrate and fire (leaky-integrate-and-fire LIF) 
'''

# Initialize vairables.
T = 50 # total time to simulate (msec)
dt = 0.125 # simulation time step (msec)
time = arange(0, T+dt, dt) # time array
trest = 0 # initial refractory time
Vm = zeros(len(time)) # potential (V) trace over time
Rm = 1 # resistance (kOhm)
Cm = 10 # capacitance (uF)
taum = Rm*Cm # time constant (msec)
tauref = 4 # refractory period (msec)
Vth = 1 # spike threshold (V)
Vspike = 0.5 # spike delta (V)

# Input stimulus
I = 1.5 # input current (A)

# Iterate over each time step.
for i, t in enumerate(time):
   if t > trest:
       Vm[i] = Vm[i-1] + (-Vm[i-1] + I*Rm) / taum * dt
   if Vm[i] >= Vth:
       Vm[i] += Vspike
trest = t + tauref

# Plot.
plt.plot(time, Vm)
plt.title('Leaky Integrate-and-Fire Example')
plt.ylabel('Membrane Potential (V)')
plt.xlabel('Time (msec)')
plt.ylim([0,2])
plt.show()
